core/util.bzl

```diff
@@ -47,6 +47,70 @@ def cp(repository_ctx, src, dest = None):
 
     return dest
 
+def cp_many(repository_ctx, srcs):
+    """Copy the given files into the external repository root.
+
+    If possible this is done by invoking tar from an external shell process.
+    This should be more efficient than writing files with repository_ctx.file
+    and preserving their permissions by invoking chmod on every file.
+
+    Args:
+      repository_ctx: The repository context of the current repository rule.
+      srcs: A list of source files to copy.
+
+    Returns:
+      A list of destination paths for the files copied into the repository.
+    """
+    tar_list = []
+    repository_files = []
+
+    # Tar will run with outputBase as its working directory.
+    repository_dir = repository_ctx.execute(["pwd"]).stdout.strip("\n")
+    output_base = repository_dir.removesuffix("external/" + repository_ctx.name)
+
+    # Fallback to cp if outputBase cannot be determined.
+    if output_base == repository_dir:
+        return [cp(repository_ctx, src) for src in srcs]
+
+    for src in srcs:
+        abs_path = str(repository_ctx.path(src))
+        rel_path = abs_path.removeprefix(output_base)
+        if rel_path != abs_path:
+            # This file can be copied with tar.
+            tar_list.append((src, rel_path))
+        else:
+            # If file is not under outputBase, fallback to cp.
+            # This can happen if Bazel is running in Bazel and
+            # the file was generated by the outer Bazel.
+            repository_files.append(cp(repository_ctx, src))
+
+    # Write a list of paths to be included in the tar archive.
+    tar_list_content = "\n".join([p for (_, p) in tar_list])
+    tar_list_filename = "tarlist{}".format(hash(tar_list_content))
+    repository_ctx.file(tar_list_filename, tar_list_content)
+
+    # Invoke external shell process to copy files with tar.
+    tar_result = repository_ctx.execute(["sh", "-c", """
```

Comment:
[`execute`](https://bazel.build/rules/lib/repository_ctx#execute) has a `working_directory` parameter, so the `cd` shouldn't be necessary.

Reply:
Nice, will change.